node/core/provisioner/src/disputes/with_staging_api/mod.rs

```diff
@@ -0,0 +1,402 @@
+	let mut total_votes_len = 0;
+	let mut result = BTreeMap::new();
+	while !disputes.is_empty() {
+		let batch_size = std::cmp::min(BATCH_SIZE, disputes.len());
+		let batch = Vec::from_iter(disputes.drain(0..batch_size));
+
+		// Filter votes which are already onchain
+		let votes = super::request_votes(sender, batch)
+			.await
+			.into_iter()
+			.map(|(session_index, candidate_hash, mut votes)| {
+				let onchain_state =
+					if let Some(onchain_state) = onchain.get(&(session_index, candidate_hash)) {
+						onchain_state
+					} else {
+						// onchain knows nothing about this dispute - add all votes
+						return (session_index, candidate_hash, votes)
+					};
+
+				votes.valid.retain(|v| is_vote_worth_to_keep(v, onchain_state));
+				votes.invalid.retain(|v| is_vote_worth_to_keep(v, onchain_state));
+				(session_index, candidate_hash, votes)
+			})
+			.collect::<Vec<_>>();
+
+		// Check if votes are within the limit
+		for (session_index, candidate_hash, selected_votes) in votes {
+			let votes_len = selected_votes.valid.len() + selected_votes.invalid.len();
+			if votes_len + total_votes_len > MAX_DISPUTE_VOTES_FORWARDED_TO_RUNTIME {
+				// we are done - no more votes can be added
+				return result
+			}
+			result.insert((session_index, candidate_hash), selected_votes);
+			total_votes_len += votes_len
+		}
+	}
+
+	result
+}
+
+/// Contains disputes by partitions. Check the field comments for further details.
+#[derive(Default)]
+pub(crate) struct PartitionedDisputes {
+	/// Concluded and inactive disputes which are completely unknown for the Runtime.
+	/// Hopefully this should never happen.
+	/// Will be sent to the Runtime with FIRST priority.
+	pub inactive_unknown_onchain: Vec<(SessionIndex, CandidateHash)>,
+	/// Active disputes completely unknown onchain.
+	/// Will be sent to the Runtime with SECOND priority.
+	pub active_unknown_onchain: Vec<(SessionIndex, CandidateHash)>,
+	/// Active disputes unconcluded onchain.
+	/// Will be sent to the Runtime with THIRD priority.
+	pub active_unconcluded_onchain: Vec<(SessionIndex, CandidateHash)>,
+	/// Active disputes concluded onchain. New votes are not that important for
+	/// this partition.
+	/// Will be sent to the Runtime with FOURTH priority.
+	pub active_concluded_onchain: Vec<(SessionIndex, CandidateHash)>,
+	/// Inactive disputes which are known onchain. These are not
+	/// interesting and won't be sent to the Runtime.
+	pub inactive_known_onchain: Vec<(SessionIndex, CandidateHash)>,
+}
+
+impl PartitionedDisputes {
+	fn new() -> PartitionedDisputes {
+		Default::default()
+	}
+
+	fn into_iter(self) -> impl Iterator<Item = (SessionIndex, CandidateHash)> {
+		self.inactive_unknown_onchain
+			.into_iter()
+			.chain(self.active_unknown_onchain.into_iter())
+			.chain(self.active_unconcluded_onchain.into_iter())
+			.chain(self.active_concluded_onchain.into_iter())
+		// inactive_known_onchain is dropped on purpose
+	}
+}
+
+fn secs_since_epoch() -> Timestamp {
```

Comment:
This is more or less copy-pasted from [dispute-coordinator](https://github.com/paritytech/polkadot/blob/566a403be1277612f413136d9febe5efaabcf417/node/core/dispute-coordinator/src/status.rs#L142)

Can we move the `SystemClock` to host primitives and reuse the code? 
I'm also wondering why it is a struct instead of a free standing function returning the current time? It doesn't keep any state.

Reply:
`SystemTime` can't go in primitives because you can't access the time from the runtime Wasm execution environment. It wouldn't compile to Wasm. I don't have an opinion on the rest, though - not sure what struct you're referring to but `SystemTime` comes from rust std.

Reply:
It could go in node-primitives or subsystem-util or any other grab-bag. I'm not really opinionated on the refactoring but if this is popping up in a bunch of places, yeah, let's extract it somewhere.

Reply:
I've created a task for this: https://github.com/paritytech/polkadot/issues/5954